src/java.base/share/classes/java/text/CompactNumberFormat.java

```diff
@@ -2354,10 +2354,14 @@ public void setParseBigDecimal(boolean newValue) {
     @Override
     public boolean equals(Object obj) {
 
-        if (!super.equals(obj)) {
+        if (!super.equals(obj)) { // super does null and class checks
             return false;
         }
 
+        if (obj == this) { // disambiguate super equality from ref equality
+            return true;
+        }
+
```

Comment:
True, there's no simple way to avoid the double check in either the equal or not equal case.
And I don't know enough about the Hotspot compiler to know whether it would spot the duplication and optimize it.
In many `equals` methods, the test for self equality is frequently the first line.
Is fine as is.

Reply:
Re-arranged the checks in f1ce3a1039a. You are right, this way is more conventional. As a bonus, we can dispense with the comment.